### ISSUE TITLE
Mangled tmst and many dupes

### ISSUE THREAD

**Issue by lthiery** [helium/packet-purchaser#29]
I configured [a virtual device](https://github.com/helium/virtual-lorawan-device) (modified to output tmst) and connected it to an unasserted synced miner. I configured Packet Purchaser to route Join Requests for my test device to a test endpoint, configured with [lorawan-sniffer](git@github.com:helium/lorawan-sniffer.git).

I allowed the device to send two Join Requests before stopping the program. The `tmst` line is added and printed when the UDP packet is dispatched to Miner.

```~/virtual-lorawan-device
[2021-08-16T08:33:55Z INFO  virtual_lorawan_device::virtual_device::udp_radio] Connecting to server 127.0.0.1:1680 from 0.0.0.0:0
tmst 908
[2021-08-16T08:34:01Z WARN  virtual_lorawan_device::virtual_device] No Join Accept Received
tmst 6101071
```

[The full Miner logs are attached](https://github.com/helium/packet-purchaser/files/6990971/miner_packet_purchaser_test.log), but in summary, things appear to go well:

The first packet is transacted
```
2021-08-16 08:42:58.712 7 [info] <0.1547.0>@miner_lora:handle_udp_packet:366 PUSH_DATA [{<<"rxpk">>,[[{<<"chan">>,0},{<<"codr">>,<<"4/5">>},{<<"data">>,<<"AAAAAQAADCUA7qAAAAH///+bg3NaQWY=">>},{<<"datr">>,<<"SF10BW125">>},{<<"freq">>,904.7},{<<"lsnr">>,5.5},{<<"modu">>,<<"LORA">>},{<<"rfch">>,0},{<<"rssi">>,-112},{<<"size">>,23},{<<"stat">>,1},{<<"tmst">>,274}]]}] from 72623859790382856 on 37061
2021-08-16 08:42:58.712 7 [notice] <0.1547.0>@miner_lora:handle_packets:529 Routing {eui,18446742978492932334,10427768277893120}
2021-08-16 08:42:58.713 7 [notice] <0.1521.0>@blockchain_state_channels_client:handle_cast:156 failed to find router for join packet with routing information {eui,18446742978492932334,10427768277893120}:eui_not_matched, trying default routers
2021-08-16 08:42:58.713 7 [info] <0.1521.0>@blockchain_state_channels_client:handle_packet:356 handle_packet #packet_pb{oui=0,type=lorawan,payload=<<0,0,0,1,0,0,12,37,0,238,160,0,0,1,255,255,255,155,131,115,90,65,102>>,timestamp=274,signal_strength=-112,frequency=904.7,datarate=<<"SF10BW125">>,snr=5.5,routing=#routing_information_pb{data={eui,{eui_pb,18446742978492932334,10427768277893120}}},rx2_window=undefined} to ["/p2p/11w77YQLhgUt8HUJrMtntGGr97RyXmot1ofs5Ct2ELTmbFoYsQa"]
2021-08-16 08:43:00.490 7 [info] <0.28527.2>@blockchain_state_channel_handler:handle_data:153 sc_handler client got banner, sc_id: <<142,36,36,142,174,217,229,225,23,167,76,150,9,66,150,4,123,16,156,46,168,37,197,189,231,102,137,177,13,198,222,223>>
2021-08-16 08:43:00.510 7 [info] <0.1521.0>@blockchain_state_channels_client:maybe_send_packets:741 valid banner for "/p2p/11w77YQLhgUt8HUJrMtntGGr97RyXmot1ofs5Ct2ELTmbFoYsQa", sending 1 packets
```

The second packet is transacted:
```
2021-08-16 08:43:04.815 7 [info] <0.1547.0>@miner_lora:handle_udp_packet:366 PUSH_DATA [{<<"rxpk">>,[[{<<"chan">>,0},{<<"codr">>,<<"4/5">>},{<<"data">>,<<"AAAAAQAADCUA7qAAAAH////FfKjoTU4=">>},{<<"datr">>,<<"SF10BW125">>},{<<"freq">>,904.1},{<<"lsnr">>,5.5},{<<"modu">>,<<"LORA">>},{<<"rfch">>,0},{<<"rssi">>,-112},{<<"size">>,23},{<<"stat">>,1},{<<"tmst">>,6103072}]]}] from 72623859790382856 on 37061
2021-08-16 08:43:04.816 7 [notice] <0.1547.0>@miner_lora:handle_packets:529 Routing {eui,18446742978492932334,10427768277893120}
2021-08-16 08:43:04.816 7 [notice] <0.1521.0>@blockchain_state_channels_client:handle_cast:156 failed to find router for join packet with routing information {eui,18446742978492932334,10427768277893120}:eui_not_matched, trying default routers
2021-08-16 08:43:04.816 7 [info] <0.1521.0>@blockchain_state_channels_client:handle_packet:356 handle_packet #packet_pb{oui=0,type=lorawan,payload=<<0,0,0,1,0,0,12,37,0,238,160,0,0,1,255,255,255,197,124,168,232,77,78>>,timestamp=6103072,signal_strength=-112,frequency=904.1,datarate=<<"SF10BW125">>,snr=5.5,routing=#routing_information_pb{data={eui,{eui_pb,18446742978492932334,10427768277893120}}},rx2_window=undefined} to ["/p2p/11w77YQLhgUt8HUJrMtntGGr97RyXmot1ofs5Ct2ELTmbFoYsQa"]
2021-08-16 08:43:04.819 7 [info] <0.1521.0>@blockchain_state_channels_client:send_offer:665 OfferMsg1: {blockchain_state_channel_offer_v1_pb,{routing_information_pb,{eui,{eui_pb,18446742978492932334,10427768277893120}}},<<252,30,244,157,170,213,181,118,87,171,110,170,255,204,208,251,46,52,247,167,201,94,37,44,132,46,149,107,81,151,75,64>>,23,0,<<0,238,238,0,211,83,133,110,218,196,47,176,217,1,105,251,23,236,205,56,14,61,38,38,162,122,83,153,239,201,252,104,117>>,<<48,68,2,32,47,134,119,235,107,248,247,208,178,104,34,254,232,227,21,82,72,43,144,223,38,46,27,134,62,143,64,199,135,191,93,107,2,32,67,228,47,120,94,67,97,228,142,161,32,49,184,134,196,211,65,78,145,57,217,45,8,243,241,91,178,212,253,8,136,1>>,'US915'}
2021-08-16 08:43:04.819 7 [info] <0.1521.0>@blockchain_state_channel_handler:send_offer:70 sending offer: {blockchain_state_channel_offer_v1_pb,{routing_information_pb,{eui,{eui_pb,18446742978492932334,10427768277893120}}},<<252,30,244,157,170,213,181,118,87,171,110,170,255,204,208,251,46,52,247,167,201,94,37,44,132,46,149,107,81,151,75,64>>,23,0,<<0,238,238,0,211,83,133,110,218,196,47,176,217,1,105,251,23,236,205,56,14,61,38,38,162,122,83,153,239,201,252,104,117>>,<<48,68,2,32,47,134,119,235,107,248,247,208,178,104,34,254,232,227,21,82,72,43,144,223,38,46,27,134,62,143,64,199,135,191,93,107,2,32,67,228,47,120,94,67,97,228,142,161,32,49,184,134,196,211,65,78,145,57,217,45,8,243,241,91,178,212,253,8,136,1>>,'US915'}, pid: <0.28527.2>
2021-08-16 08:43:05.178 7 [info] <0.28527.2>@blockchain_state_channel_handler:handle_data:165 sc_handler client got purchase, sc_id: <<239,212,148,161,200,235,144,90,145,142,235,136,208,238,13,36,54,255,217,161,210,159,165,80,205,163,47,124,192,131,194,251>>
```

lorawan-sniffer was waiting at the endpoint `lorawan-sniffer -- --in-port 1680 --enable-tmst`

Here is the complete output:

```
[2021-08-16 08:43:00.891]  JoinRequest	904.7 MHz 	SF10BW125	RSSI: -112	LSNR: 5.5
	tmst: 1,310,775,410
	  AppEui: 00250C0000010000 DevEui: FFFFFF010000A0EE DevNonce: 839B
[2021-08-16 08:43:02.893]  JoinRequest	904.7 MHz 	SF10BW125	RSSI: -112	LSNR: 5.5
	tmst: 1,310,775,410
	  AppEui: 00250C0000010000 DevEui: FFFFFF010000A0EE DevNonce: 839B
[2021-08-16 08:43:04.894]  JoinRequest	904.7 MHz 	SF10BW125	RSSI: -112	LSNR: 5.5
	tmst: 1,310,775,410
	  AppEui: 00250C0000010000 DevEui: FFFFFF010000A0EE DevNonce: 839B
[2021-08-16 08:43:05.292]  JoinRequest	904.1 MHz 	SF10BW125	RSSI: -112	LSNR: 5.5
	tmst: 1,310,779,720
	  AppEui: 00250C0000010000 DevEui: FFFFFF010000A0EE DevNonce: 7CC5
[2021-08-16 08:43:06.895]  JoinRequest	904.7 MHz 	SF10BW125	RSSI: -112	LSNR: 5.5
	tmst: 1,310,775,410
	  AppEui: 00250C0000010000 DevEui: FFFFFF010000A0EE DevNonce: 839B
[2021-08-16 08:43:07.293]  JoinRequest	904.1 MHz 	SF10BW125	RSSI: -112	LSNR: 5.5
	tmst: 1,310,779,720
	  AppEui: 00250C0000010000 DevEui: FFFFFF010000A0EE DevNonce: 7CC5
[2021-08-16 08:43:08.896]  JoinRequest	904.7 MHz 	SF10BW125	RSSI: -112	LSNR: 5.5
	tmst: 1,310,775,410
	  AppEui: 00250C0000010000 DevEui: FFFFFF010000A0EE DevNonce: 839B
```

The output here has two issues.

**The timestamps are all off**
Recall that `tmst` is expressed in ~~ns~~ μs. Therefore, the virtual device emits the frames: `(6101071 - 908)/10**6 = 6.100163 seconds` apart which is a JoinRequest, and two expired RX windows followed by an immediate second JoinRequest. Note that these timestamps are parsed properly in the Miner logs. Meanwhile, the reported timestamps to the lorawan-sniffer by packet-purchaser are incredible: `(1310779720 - 1310775410)/10**6 = 0.00431 seconds` apart.

**The same two packets appear many times**
You can determine this by the unique DevNonce where 839B appears 5 times and 7CC5 appears 2 times. A single Miner is being used with a virtual device here, so there is no chance that these are being reported by another Miner.

**Comment by Vagabond** [comment]
But I don't think we change TMST at all, other than to keep it bounded within a 32 bit number. Isn't that the LNS' job?

**Comment by Vagabond** [comment]
https://github.com/helium/packet-purchaser/blob/811e9a3202753106f8af3301c197554f9ffdd85c/src/state_channels/pp_sc_packet_handler.erl#L83 is the only place that seems to change TMST at all.

**Comment by Vagabond** [comment]
Additionally, according to https://www.thethingsnetwork.org/forum/uploads/default/original/1X/4fbda86583605f4aa24dcedaab874ca5a1572825.pdf page 11, the tmst is in microseconds, not nanoseconds.

**Comment by Vagabond** [comment]
The duplicate packet issue does look worth investigating.

**Comment by Vagabond** [comment]
Do you know if the same packet came in several times from the same udp source port/gwid or if those changed?

**Comment by Vagabond** [comment]
Is your lorawan-sniffer sending PULL_ACK messages? I think the packet purchaser is not very tolerant of missing PULL_ACK messages and it will close/reconnect the udp connection (which may trigger resends).

**Comment by Vagabond** [comment]
Yeah, not sending a PUSH_ACK message in response to a PUSH_DATA udp packet will trigger a re-send: https://github.com/helium/packet-purchaser/blob/master/src/pp_udp_worker.erl#L128

This is not correct, and I've filed #30 to address that.

**Comment by Vagabond** [comment]
What is the explanation for the tmst going from 908 -> 274 -> 1,310,775,410

**Comment by lthiery** [comment]
> Yeah, not sending a PUSH_ACK message in response to a PUSH_DATA udp packet will trigger a re-send: https://github.com/helium/packet-purchaser/blob/master/src/pp_udp_worker.erl#L128
> 
> This is not correct, and I've filed #30 to address that.

That seems like a good explanation then and makes it less bothersome. LoRaWAN sniffer is just receiving the packets and doing nothing else about the protocol

**Comment by lthiery** [comment]
> What is the explanation for the tmst going from 908 -> 274 -> 1,310,775,410

My bad on this. The Miner logs are from a different tracing session, but they did match the output from the virtual device is the point

**Comment by michaeldjeffrey** [comment]
Fixed by #31
